## Assignment: $k$ Means Clustering

## **Do two questions.**

`! git clone https://www.github.com/DS3001/kmc`

In [1]:
! git clone https://github.com/bwillow1222/kmc

Cloning into 'kmc'...
remote: Enumerating objects: 25, done.
remote: Counting objects: 100% (4/4), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 25 (delta 2), reused 1 (delta 1), pack-reused 21
Receiving objects: 100% (25/25), 5.04 MiB | 19.20 MiB/s, done.
Resolving deltas: 100% (4/4), done.


In [2]:
import pandas as pd

**Q1.** This question is a case study for $k$ means clustering.

1. Load the `airbnb_hw.csv` data. Clean `Price` along with `Beds`, `Number of Reviews`, and `Review Scores Rating`.


In [52]:
df = pd.read_csv('/content/kmc/data/airbnb_hw.csv')

# Make a df of variables of interest
X = df.loc[:,['Price','Beds','Number Of Reviews','Review Scores Rating']]

# Print summary stats of new df
print(X.shape)
X.describe()

(30478, 4)


Beds  Number Of Reviews  Review Scores Rating
count  30393.000000       30478.000000          22155.000000
mean       1.530089          12.018735             91.993230
std        1.015359          21.980703              8.850373
min        0.000000           0.000000             20.000000
25%        1.000000           0.000000             89.000000
50%        1.000000           3.000000             94.000000
75%        2.000000          13.000000            100.000000
max       16.000000         257.000000            100.000000

In [53]:
# Clean Price
X['Price'] = X['Price'].str.replace(',','')
X['Price'] = pd.to_numeric(X['Price'],errors='coerce')

In [54]:
# Clean Beds
X['Beds'].unique()

array([ 1.,  3.,  2.,  4.,  5., nan,  6., 10.,  7.,  8., 12.,  0., 16.,
        9., 11.])

In [55]:
# Since there a missing values for beds, we are going to assume that there is 1 bed on the property
X['Beds'] = X['Beds'].fillna(1)

In [56]:
# Clean Number Of Reviews
X['Number Of Reviews'].unique()
X['Number Of Reviews'].value_counts()


0      7814
1      3572
2      2457
3      1764
4      1382
       ... 
216       1
191       1
213       1
178       1
130       1
Name: Number Of Reviews, Length: 205, dtype: int64

In [57]:
# Clean Review Scores Rating
X['Review Scores Rating'].unique()

array([ nan,  96., 100.,  94.,  90.,  98.,  93.,  91.,  97.,  95.,  99.,
        85.,  86.,  80.,  88.,  92.,  89.,  82.,  87.,  81.,  76.,  78.,
        83.,  66.,  84.,  72.,  79.,  60.,  40.,  62.,  74.,  77.,  50.,
        71.,  75.,  73.,  69.,  65.,  68.,  70.,  67.,  64.,  20.,  57.,
        58.,  43.,  63.,  55.,  47.,  53.,  49.,  30.])

In [58]:
# Don't want properties with missing review or scores so drop them
X = X.dropna()
X.describe()

Price          Beds  Number Of Reviews  Review Scores Rating
count  22155.000000  22155.000000       22155.000000          22155.000000
mean     154.787633      1.556985          16.505439             91.993230
std      148.836621      1.043273          24.308241              8.850373
min       10.000000      0.000000           1.000000             20.000000
25%       85.000000      1.000000           2.000000             89.000000
50%      125.000000      1.000000           7.000000             94.000000
75%      190.000000      2.000000          20.000000            100.000000
max    10000.000000     16.000000         257.000000            100.000000

2. Maxmin normalize the data and remove any `nan`'s (`KMeans` from `sklearn` doesn't accept `nan` input).


In [62]:
def maxmin(x):
    u = (x-min(x))/(max(x)-min(x))
    return u

X2 = X.drop('Price', axis=1)
X2 = X.apply(maxmin)

3. Use `sklearn`'s `KMeans` module to cluster the data by `Beds`, `Number of Reviews`, and `Review Scores Rating` for `k=6`.


In [63]:
from sklearn.cluster import KMeans # Import kmc
import matplotlib.pyplot as plt


model = KMeans(n_clusters=6, max_iter=300, n_init = 10, random_state=0) # Create a model for
model = model.fit(X2) # Fit the emodel
X2['cluster'] = model.labels_

In [64]:
X2.describe()

Price          Beds  Number Of Reviews  Review Scores Rating  \
count  22155.000000  22155.000000       22155.000000          22155.000000   
mean       0.014493      0.097312           0.060568              0.899915   
std        0.014899      0.065205           0.094954              0.110630   
min        0.000000      0.000000           0.000000              0.000000   
25%        0.007508      0.062500           0.003906              0.862500   
50%        0.011512      0.062500           0.023438              0.925000   
75%        0.018018      0.125000           0.074219              1.000000   
max        1.000000      1.000000           1.000000              1.000000   

            cluster  
count  22155.000000  
mean       2.360144  
std        1.976268  
min        0.000000  
25%        0.000000  
50%        3.000000  
75%        4.000000  
max        5.000000

4. Use `seaborn`'s `.pairplot()` to make a grid of scatterplots that show how the clustering is carried out in multiple dimensions.


5. Use `.groupby` and `.describe` to compute the average price for each cluster. Which clusters have the highest rental prices?


6. Use a scree plot to pick the number of clusters and repeat steps 4 and 5.

**Q2.** This is a question about $k$ means clustering. We want to investigate how adjusting the "noisiness" of the data impacts the quality of the algorithm and the difficulty of picking $k$.

1. Run the code below, which creates four datasets: `df0_125`, `df0_25`, `df0_5`, `df1_0`, and `df2_0`. Each data set is created by increasing the amount of `noise` (standard deviation) around the cluster centers, from `0.125` to `0.25` to `0.5` to `1.0` to `2.0`.

```
import numpy as np
import pandas as pd

def createData(noise,N=50):
    np.random.seed(100) # Set the seed for replicability
    # Generate (x1,x2,g) triples:
    X1 = np.array([np.random.normal(1,noise,N),np.random.normal(1,noise,N)])
    X2 = np.array([np.random.normal(3,noise,N),np.random.normal(2,noise,N)])
    X3 = np.array([np.random.normal(5,noise,N),np.random.normal(3,noise,N)])
    # Concatenate into one data frame
    gdf1 = pd.DataFrame({'x1':X1[0,:],'x2':X1[1,:],'group':'a'})
    gdf2 = pd.DataFrame({'x1':X2[0,:],'x2':X2[1,:],'group':'b'})
    gdf3 = pd.DataFrame({'x1':X3[0,:],'x2':X3[1,:],'group':'c'})
    df = pd.concat([gdf1,gdf2,gdf3],axis=0)
    return df

df0_125 = createData(0.125)
df0_25 = createData(0.25)
df0_5 = createData(0.5)
df1_0 = createData(1.0)
df2_0 = createData(2.0)
```

2. Make scatterplots of the $(X1,X2)$ points by group for each of the datasets. As the `noise` goes up from 0.125 to 2.0, what happens to the visual distinctness of the clusters?
3. Create a scree plot for each of the datasets. Describe how the level of `noise` affects the scree plot (particularly the presence of a clear "elbow") and your ability to definitively select a $k$.
4. Explain the intuition of the elbow, using this numerical simulation as an example.

**Q3.** We looked at computer vision with $k$NN in a previous question. Can $k$ means clustering correctly group digits, even if we don't know which symbols are which?

1. To load the data, run the following code in a chunk:
```
from keras.datasets import mnist
df = mnist.load_data('minst.db')
train,test = df
X_train, y_train = train
X_test, y_test = test
```
The `y_test` and `y_train` vectors, for each index `i`, tell you want number is written in the corresponding index in `X_train[i]` and `X_test[i]`. The value of `X_train[i]` and `X_test[i]`, however, is a 28$\times$28 array whose entries contain values between 0 and 256. Each element of the matrix is essentially a "pixel" and the matrix encodes a representation of a number. To visualize this, run the following code to see the first ten numbers:
```
import matplotlib.pyplot as plt
import numpy as np
np.set_printoptions(edgeitems=30, linewidth=100000)
for i in range(5):
    print(y_test[i],'\n') # Print the label
    print(X_test[i],'\n') # Print the matrix of values
    plt.contourf(np.rot90(X_test[i].transpose())) # Make a contour plot of the matrix values
    plt.show()
```
OK, those are the data: Labels attached to handwritten digits encoded as a matrix.

2. What is the shape of `X_train` and `X_test`? What is the shape of `X_train[i]` and `X_test[i]` for each index `i`? What is the shape of `y_train` and `y_test`?
3. Use Numpy's `.reshape()` method to covert the training and testing data from a matrix into an vector of features. So, `X_test[index].reshape((1,784))` will convert the $index$-th element of `X_test` into a $28\times 28=784$-length row vector of values, rather than a matrix. Turn `X_train` into an $N \times 784$ matrix $X$ that is suitable for scikit-learn's kNN classifier where $N$ is the number of observations and $784=28*28$ (you could use, for example, a `for` loop).
4. Use $k$ means clustering on the reshaped `X_test` data with `k=10`.  
5. Cross tabulate the cluster assignments with the true labels for the test set values. How good is the correspondence? What proportion of digits are clustered correctly? Which digits are the hardest to distinguish from one another? Can $k$MC recover the latent digits 0 to 9, without even knowing what those digits were?
6. If you use a scree plot to determine the number of clusters $k$, does it pick 10 (the true number of digits), or not? If it fails to pick $k=10$, which digits does it tend to combine into the same classification?